File: kits/php/src/StrictArgs.php
<?php

declare(strict_types=1);

namespace IronMcp;

/**
 * The 3-state rule + the advertise stamp. No SDK import; pure. The schema is authoritative, and a
 * guard that bricks what it cannot read is worse than the bug it prevents.
 */
final class StrictArgs
{
    /**
     * @param array<string, mixed>|null $schema a JSON Schema object as a PHP array
     * @param array<string, mixed>|null $args
     * @return array{ok: true}|array{ok: false, unknown: list<string>, accepted: list<string>, message: string}
     */
    public static function check(?array $schema, ?array $args, string $toolName = 'tool', string $reconnectHint = Messages::DEFAULT_RECONNECT_HINT): array
    {
        if (!self::isEnforced($schema)) {
            return ['ok' => true]; // states 1 (unintrospectable) & 3 (opted open)
        }
        // properties may be a stdClass when an SDK normalised an empty {} — cast to array.
        $accepted = array_keys((array) ($schema['properties'] ?? []));
        $acceptedSet = array_fill_keys($accepted, true);
        $unknown = array_values(array_filter(
            array_keys($args ?? []),
            static fn ($k) => !isset($acceptedSet[$k]),
        ));
        sort($unknown);
        if ($unknown === []) {
            return ['ok' => true];
        }

        return [
            'ok' => false,
            'unknown' => $unknown,
            'accepted' => $accepted,
            'message' => Messages::unknownArgs($toolName, $unknown, $accepted, $reconnectHint),
        ];
    }

    /**
     * @param array<string, mixed>|null $schema
     * @return array<string, mixed>|null
     */
    public static function stampClosed(?array $schema): ?array
    {
        if (!self::isEnforced($schema)) {
            return $schema;
        }
        $schema['additionalProperties'] = false;

        return $schema;
    }

    /** True only in state 2: an INTROSPECTABLE properties map is present and the tool is not opted open. */
    private static function isEnforced(?array $schema): bool
    {
        if ($schema === null || !array_key_exists('properties', $schema)) {
            return false;
        }
        if (($schema['additionalProperties'] ?? null) === true) {
            return false; // state 3: opted open
        }

        return self::propertiesIntrospectable($schema['properties']);
    }

    /**
     * A `properties` value is introspectable as a name map when it is an object shape: an empty
     * array (`{}`), an associative array, or a stdClass (an SDK-normalised empty `{}`). A malformed
     * schema — `properties` given as a string or a populated JSON list — is NOT introspectable, and
     * a guard that bricks what it cannot read is worse than the bug it prevents: treat it as state 1
     * (permissive), never refuse every argument.
     */
    private static function propertiesIntrospectable(mixed $properties): bool
    {
        if ($properties instanceof \stdClass) {
            return true;
        }
        if (\is_array($properties)) {
            return $properties === [] || !array_is_list($properties);
        }

        return false;
    }
}

File: kits/dart/lib/src/strict_args.dart
import 'messages.dart';

/// The outcome of a strict-args check: either accepted, or a refusal carrying
/// the offending keys and the ready-made message.
class StrictResult {
  const StrictResult.ok()
      : ok = true,
        unknown = const [],
        accepted = const [],
        message = null;

  const StrictResult.refused({
    required this.unknown,
    required this.accepted,
    required this.message,
  }) : ok = false;

  final bool ok;
  final List<String> unknown;
  final List<String> accepted;
  final String? message;
}

/// The 3-state rule + the advertise stamp. No SDK import; pure. The schema is
/// authoritative, and a guard that bricks what it cannot read is worse than the
/// bug it prevents — so it enforces ONLY when the schema is introspectable and
/// not explicitly opted open.
///
/// - State 1 (unintrospectable: no `properties`)  -> permissive.
/// - State 2 (has `properties`, not opted open)    -> enforce; refuse unknowns.
/// - State 3 (`additionalProperties: true`)        -> permissive (opted out).
class StrictArgs {
  StrictArgs._();

  static StrictResult check(
    Map<String, Object?>? schema,
    Map<String, Object?>? args, {
    String toolName = 'tool',
    String reconnectHint = Messages.defaultReconnectHint,
  }) {
    if (!_isEnforced(schema)) {
      return const StrictResult.ok(); // states 1 & 3
    }
    final props = schema!['properties'];
    final accepted = props is Map
        ? props.keys.map((k) => k.toString()).toList()
        : <String>[];
    final acceptedSet = accepted.toSet();
    final unknown = (args ?? const <String, Object?>{})
        .keys
        .map((k) => k.toString())
        .where((k) => !acceptedSet.contains(k))
        .toList()
      ..sort();
    if (unknown.isEmpty) return const StrictResult.ok();

    return StrictResult.refused(
      unknown: unknown,
      accepted: accepted,
      message: Messages.unknownArgs(
        toolName,
        unknown,
        accepted,
        reconnectHint: reconnectHint,
      ),
    );
  }

  /// Return a copy of [schema] with `additionalProperties: false` when it is
  /// enforced (state 2); otherwise return [schema] unchanged (same reference),
  /// so an opted-open or unintrospectable schema is never quietly closed.
  static Map<String, Object?>? stampClosed(Map<String, Object?>? schema) {
    if (!_isEnforced(schema)) return schema;
    return {...schema!, 'additionalProperties': false};
  }

  /// True only in state 2: `properties` present (even empty) and not opted open.
  static bool _isEnforced(Map<String, Object?>? schema) {
    return schema != null &&
        schema.containsKey('properties') &&
        schema['additionalProperties'] != true;
  }
}
